README.md
<p align=center>
  <a href="https://github.com/TimoKunze/DateTimeControls/releases/tag/1.5.4">
    <img alt="Release 1.5.4 Release" src="https://img.shields.io/badge/release-1.5.4-0688CB.svg">
  </a>
  <a href="https://github.com/TimoKunze/DateTimeControls/releases">
    <img alt="Download DateTimeControls" src="https://img.shields.io/badge/download-latest-0688CB.svg">
  </a>
  <a href="https://github.com/TimoKunze/DateTimeControls/blob/master/LICENSE">
    <img alt="License" src="https://img.shields.io/badge/license-MIT-0688CB.svg">
  </a>
  <a href="https://www.paypal.com/xclick/business=TKunze71216%40gmx.de&item_name=DateTimeControls&no_shipping=1&tax=0&currency_code=EUR">
    <img alt="Donate" src="https://img.shields.io/badge/%24-donate-E44E4A.svg">
  </a>
</p>

# DateTimeControls
An ActiveX control library for Visual Basic 6 that contains various kinds of date pickers and calendars.

I've developed this ActiveX control in 2009 and did update it on a regular basis until 2016. Currently I have little interest to maintain this project any longer, but I think the code might be of some use to others.

# Before you make changes
If you make changes to the code and deploy the binary, keep in mind that ActiveX controls are COM components and therefore should stay binary compatible as long as you don't change the COM object's, i.e. the ActiveX control's public class name and GUIDs. Otherwise people using these components are likely to end up in the famous COM hell.

# Dependencies
You will need the [Microsoft Windows 10 SDK](https://developer.microsoft.com/en-us/windows/downloads/windows-10-sdk), ATL, and [WTL 10](https://sourceforge.net/projects/wtl/).

## Required Change to ATL
Some version of ATL have a bug in ```AtlIPersistPropertyBag_Load``` which causes crashes. In the file atlcom.h search for ```AtlIPersistPropertyBag_Load```. Inside the implementation of this function search for ```HRESULT hr = pPropBag->Read(pMap[i].szDesc, &var, pErrorLog);```. Make sure that there is this code before this line: ```var.pdispVal = NULL;```.

CHANGELOG
Version 1.5.4 - 27/01/2018

- Upgraded to WTL 10.
- Integrated the GitHub repository into the documentation and the About dialog.



Version 1.5.3 - 09/01/2018

- Project published on GitHub.



Version 1.5.2 - 15/11/2016

- Fixed some bugs in the DateTimePicker control that had consequences on Linux/Wine only.



Version 1.5.1 - 29/01/2016

- FIX: The controls could crash during OLE drag'n'drop.
- FIX: The controls could crash the other process during OLE drag'n'drop, especially when IOLEDataObject::SetDropDescription() had been called.



Version 1.5.0 - 22/01/2016

- Added the FinishOLEDragDrop method to each control. It should not be used if not really needed. It can be used to hide the drag image if execution of the OLEDragDrop event takes a long time or if a message is displayed from within this event.
- Initial support for Windows 10.



Version 1.4.0 - 02/07/2014

- Added the XClick, and XDblClick events for support of the extended mouse buttons.
- Extended the Mouse* events to support the extended mouse buttons.
- Added the MouseWheel events for handling mouse wheel rotations.
- Reactivated data-binding capabilities. When the library is registered, it will call regtlib.exe msdatsrc.tlb before registering itself. This seems to fix the mess caused by the broken Windows Update from early 2013.
- FIX: The Font and MouseIcon properties could leak GDI resources.
- Initial support for Windows 8.1.



Version 1.3.3 - 28/03/2013

- FIX: The GetIdealSize method ignored the Appearance and BorderStyle properties.
- Disabled data-binding capabilities as they cause "Object library not registered" errors when adding the control library to the project. This problem might be related to a recent Windows Update.



Version 1.3.2 - 18/10/2012

- FIX: If the _ICalendarEvents::ContextMenu or the _IDateTimePickerEvents::CalendarContextMenu event was triggered by keyboard input, it has been fired for the wrong date.
- FIX: The ResizedControlWindow event was not fired when it should have been fired.
- Initial support for Windows 8.



Version 1.3.1 - 22/04/2012

- FIX: Setting the ICalendar::MinDate property at runtime could reset the ICalendar::MaxDate property.
- FIX: Setting the ICalendar::MaxDate property at runtime could reset the ICalendar::MinDate property.
- FIX: Setting the IDateTimePicker::MinDate property at runtime could reset the IDateTimePicker::MaxDate property.
- FIX: Setting the IDateTimePicker::MaxDate property at runtime could reset the IDateTimePicker::MinDate property.
- Reimplemented the subclassing code of the VB6 samples to make it more robust.



Version 1.3.0 - 01/11/2011

- Added the ICalendar::DetectDoubleClicks property to completely deactivate double clicks and handle them as two single clicks instead. Attention: This property defaults to False!
- Added the IDateTimePicker::DetectDoubleClicks property to completely deactivate double clicks and handle them as two single clicks instead.



Version 1.2.2 - 20/05/2011

- FIX: Disabling the controls did not move the keyboard focus to the next control in the tab order.
- FIX: The ResizedControlWindow event was fired when actually nothing had changed.



Version 1.2.1 - 26/01/2011

- FIX: ICalendar::CaretDate could return invalid values.



Version 1.2.0 - 07/01/2011

- FIX: Due do a bug in Microsoft ATL, the controls could lose keyboard focus when they shouldn't. This problem occured in MDI applications only. Many thanks to Peter Schellenbach for his help.
- Switched to Visual C++ 2010 and dropped support for Windows 2000.



Version 1.1.3 - 18/10/2010

- FIX: With version 5.x of comctl32.dll, clicking the drop-down button of the DateTimePicker control did not display the drop-down calendar, if the control did not already have the focus and the previously focussed control was e. g. a text box. This behavior was caused by a bug inside comctl32.dll.
- FIX: Calling IDateTimePicker::CloseDropDownWindow could trigger keyboard events.



Version 1.1.2 - 23/08/2010

- Added a new property page that can be used to enter Unicode texts at design time.
- FIX: Visual Basic 6 didn't raise the Validate event correctly.
- FIX: The Calendar control had random days set to bold after the control window has been created.
- FIX: The _IDateTimePickerEvents::GetCallbackFieldTextSize event sometimes was not raised when it should have been raised.
- FIX: IOLEDataObject::SetDropDescription could cause crashes in other programs.



Version 1.1.1 - 01/01/2010

- Improved support for Aero drag images.
- Improved performance.
- FIX: If the controls were placed on a Form for which the KeyPreview feature was activated, the Form_Key* handler didn't receive key strokes like Alt+PgDown or Alt+Left, if the controls had the keyboard focus.
- FIX: The 'shift' parameter that many events have, often missed the ALT key.
- Deploy (stripped) symbols with the binaries to make debugging easier. This are the pdb files in case you wonder.



Version 1.1.0 - 26/07/2009

- Added the IDateTimePicker::Value property.
- FIX: The _IDateTimePickerEvents::CurrentDateChanged was not raised when the IDateTimePicker::DateSelected property was changed.
- Fixed a potential problem with persistance.



Version 1.0.1 - 11/07/2009

- FIX: The _IDateTimePickerEvents::ContextMenu event could get raised with wrong coordinates.
- Fixed some small OLE drag'n'drop bugs.
- Massively reduced the usage of GDI handles.
- Made some internal changes that will make it easier for me to keep binary compatibility in future.



Version 1.0.0 - 28/01/2009

- First public release.

src/DispIDs.h
// DispIDs.h: Defines a DISPID for each COM property and method we're providing.

// ICalendar
// properties
#define DISPID_CAL_APPEARANCE																	1
#define DISPID_CAL_BACKCOLOR																	2
#define DISPID_CAL_BORDERSTYLE																3
#define DISPID_CAL_CALENDARCOUNT															4
#define DISPID_CAL_CALENDARGRIDCELLSELECTED										5
#define DISPID_CAL_CALENDARGRIDCELLTEXT												6
//#define DISPID_CAL_CALENDARID																	7
#define DISPID_CAL_CARETDATE																	DISPID_VALUE
#define DISPID_CAL_CARETDATETEXT															8
#define DISPID_CAL_DISABLEDEVENTS															9
#define DISPID_CAL_DONTREDRAW																	10
#define DISPID_CAL_DRAGSCROLLTIMEBASE													11
#define DISPID_CAL_ENABLED																		DISPID_ENABLED
#define DISPID_CAL_FIRSTDATE																	12
#define DISPID_CAL_FONT																				13
#define DISPID_CAL_FORECOLOR																	14
#define DISPID_CAL_HEADERTEXT																	15
#define DISPID_CAL_HIGHLIGHTTODAYDATE													16
#define DISPID_CAL_HOVERTIME																	17
#define DISPID_CAL_HWND																				18
#define DISPID_CAL_KEEPSELECTIONONNAVIGATION									19
#define DISPID_CAL_LASTDATE																		20
#define DISPID_CAL_MAXDATE																		21
#define DISPID_CAL_MAXSELECTEDDATES														22
#define DISPID_CAL_MINDATE																		23
#define DISPID_CAL_MONTHBACKCOLOR															24
#define DISPID_CAL_MOUSEICON																	25
#define DISPID_CAL_MOUSEPOINTER																26
#define DISPID_CAL_MULTISELECT																27
#define DISPID_CAL_PADDING																		28
#define DISPID_CAL_PROCESSCONTEXTMENUKEYS											29
#define DISPID_CAL_REGISTERFOROLEDRAGDROP											30
#define DISPID_CAL_RIGHTTOLEFT																31
#define DISPID_CAL_SCROLLRATE																	32
#define DISPID_CAL_SHOWTODAYDATE															33
#define DISPID_CAL_SHOWTRAILINGDATES													34
#define DISPID_CAL_SHOWWEEKNUMBERS														35
#define DISPID_CAL_STARTOFWEEK																36
#define DISPID_CAL_SUPPORTOLEDRAGIMAGES												37
#define DISPID_CAL_TITLEBACKCOLOR															38
#define DISPID_CAL_TITLEFORECOLOR															39
#define DISPID_CAL_TODAYDATE																	40
#define DISPID_CAL_TRAILINGFORECOLOR													41
#define DISPID_CAL_USEPADDING																	42
#define DISPID_CAL_USESHORTESTDAYNAMES												43
#define DISPID_CAL_USESYSTEMFONT															44
#define DISPID_CAL_VERSION																		45
#define DISPID_CAL_VIEW																				46
#define DISPID_CAL_DETECTDOUBLECLICKS													58
// fingerprint
#define DISPID_CAL_APPID																			500
#define DISPID_CAL_APPNAME																		501
#define DISPID_CAL_APPSHORTNAME																502
#define DISPID_CAL_BUILD																			503
#define DISPID_CAL_CHARSET																		504
#define DISPID_CAL_ISRELEASE																	505
#define DISPID_CAL_PROGRAMMER																	506
#define DISPID_CAL_TESTER																			507
// methods
#define DISPID_CAL_ABOUT																			DISPID_ABOUTBOX
#define DISPID_CAL_COUNTVISIBLEMONTHS													47
#define DISPID_CAL_GETMAXTODAYWIDTH														48
#define DISPID_CAL_GETMINIMUMSIZE															49
#define DISPID_CAL_GETRECTANGLE																50
#define DISPID_CAL_GETSELECTION																51
#define DISPID_CAL_HITTEST																		52
#define DISPID_CAL_LOADSETTINGSFROMFILE												53
#define DISPID_CAL_MINIMIZERECTANGLE													54
#define DISPID_CAL_REFRESH																		DISPID_REFRESH
#define DISPID_CAL_SAVESETTINGSTOFILE													55
#define DISPID_CAL_SETBOLDDATES																56
#define DISPID_CAL_SETSELECTION																57
#define DISPID_CAL_FINISHOLEDRAGDROP													59


// _ICalendarEvents
// methods
#define DISPID_CALE_CLICK																			1
#define DISPID_CALE_CONTEXTMENU																2
#define DISPID_CALE_DATEMOUSEENTER														3
#define DISPID_CALE_DATEMOUSELEAVE														4
#define DISPID_CALE_DBLCLICK																	5
#define DISPID_CALE_DESTROYEDCONTROLWINDOW										6
#define DISPID_CALE_GETBOLDDATES															7
#define DISPID_CALE_KEYDOWN																		8
#define DISPID_CALE_KEYPRESS																	9
#define DISPID_CALE_KEYUP																			10
#define DISPID_CALE_MCLICK																		11
#define DISPID_CALE_MDBLCLICK																	12
#define DISPID_CALE_MOUSEDOWN																	13
#define DISPID_CALE_MOUSEENTER																14
#define DISPID_CALE_MOUSEHOVER																15
#define DISPID_CALE_MOUSELEAVE																16
#define DISPID_CALE_MOUSEMOVE																	17
#define DISPID_CALE_MOUSEUP																		18
#define DISPID_CALE_OLEDRAGDROP																19
#define DISPID_CALE_OLEDRAGENTER															20
#define DISPID_CALE_OLEDRAGLEAVE															21
#define DISPID_CALE_OLEDRAGMOUSEMOVE													22
#define DISPID_CALE_RCLICK																		23
#define DISPID_CALE_RDBLCLICK																	24
#define DISPID_CALE_RECREATEDCONTROLWINDOW										25
#define DISPID_CALE_RESIZEDCONTROLWINDOW											26
#define DISPID_CALE_SELECTIONCHANGED													27
#define DISPID_CALE_VIEWCHANGED																28
#define DISPID_CALE_MOUSEWHEEL																29
#define DISPID_CALE_XCLICK																		30
#define DISPID_CALE_XDBLCLICK																	31


// IDateTimePicker
// properties
#define DISPID_DTP_ALLOWNULLSELECTION													1
#define DISPID_DTP_APPEARANCE																	2
#define DISPID_DTP_BORDERSTYLE																3
#define DISPID_DTP_CALENDARBACKCOLOR													4
#define DISPID_DTP_CALENDARCARETDATETEXT											5
#define DISPID_DTP_CALENDARDRAGSCROLLTIMEBASE									6
#define DISPID_DTP_CALENDARFIRSTDATE													7
#define DISPID_DTP_CALENDARFONT																8
#define DISPID_DTP_CALENDARFORECOLOR													9
#define DISPID_DTP_CALENDARGRIDCELLSELECTED										10
#define DISPID_DTP_CALENDARGRIDCELLTEXT												11
#define DISPID_DTP_CALENDARHEADERTEXT													12
#define DISPID_DTP_CALENDARHIGHLIGHTTODAYDATE									13
#define DISPID_DTP_CALENDARHOVERTIME													14
#define DISPID_DTP_CALENDARKEEPSELECTIONONNAVIGATION					15
#define DISPID_DTP_CALENDARLASTDATE														16
#define DISPID_DTP_CALENDARMONTHBACKCOLOR											17
#define DISPID_DTP_CALENDARSHOWTODAYDATE											18
#define DISPID_DTP_CALENDARSHOWTRAILINGDATES									19
#define DISPID_DTP_CALENDARSHOWWEEKNUMBERS										20
#define DISPID_DTP_CALENDARTITLEBACKCOLOR											21
#define DISPID_DTP_CALENDARTITLEFORECOLOR											22
#define DISPID_DTP_CALENDARTODAYDATE													23
#define DISPID_DTP_CALENDARTRAILINGFORECOLOR									24
#define DISPID_DTP_CALENDARUSESHORTESTDAYNAMES								25
#define DISPID_DTP_CALENDARUSESYSTEMFONT											26
#define DISPID_DTP_CALENDARVIEW																27
#define DISPID_DTP_CHECKBOXOBJECTSTATE												28
#define DISPID_DTP_CURRENTDATE																DISPID_VALUE
#define DISPID_DTP_CUSTOMDATEFORMAT														29
#define DISPID_DTP_DATEFORMAT																	30
#define DISPID_DTP_DATESELECTED																31
#define DISPID_DTP_DISABLEDEVENTS															32
#define DISPID_DTP_DONTREDRAW																	33
#define DISPID_DTP_DRAGDROPDOWNTIME														34
#define DISPID_DTP_DROPDOWNALIGNMENT													35
#define DISPID_DTP_DROPDOWNBUTTONOBJECTSTATE									36
#define DISPID_DTP_ENABLED																		DISPID_ENABLED
#define DISPID_DTP_FONT																				37
#define DISPID_DTP_HOVERTIME																	38
#define DISPID_DTP_HWND																				39
#define DISPID_DTP_HWNDCALENDAR																40
#define DISPID_DTP_HWNDEDIT																		41
#define DISPID_DTP_HWNDUPDOWN																	42
#define DISPID_DTP_IMEMODE																		43
#define DISPID_DTP_MAXDATE																		44
#define DISPID_DTP_MINDATE																		45
#define DISPID_DTP_MOUSEICON																	46
#define DISPID_DTP_MOUSEPOINTER																47
#define DISPID_DTP_PROCESSCONTEXTMENUKEYS											48
#define DISPID_DTP_REGISTERFOROLEDRAGDROP											49
#define DISPID_DTP_RIGHTTOLEFT																50
#define DISPID_DTP_STARTOFWEEK																51
#define DISPID_DTP_STYLE																			52
#define DISPID_DTP_SUPPORTOLEDRAGIMAGES												53
#define DISPID_DTP_USESYSTEMFONT															54
#define DISPID_DTP_VERSION																		55
#define DISPID_DTP_VALUE																			66
#define DISPID_DTP_DETECTDOUBLECLICKS													67
// fingerprint
#define DISPID_DTP_APPID																			500
#define DISPID_DTP_APPNAME																		501
#define DISPID_DTP_APPSHORTNAME																502
#define DISPID_DTP_BUILD																			503
#define DISPID_DTP_CHARSET																		504
#define DISPID_DTP_ISRELEASE																	505
#define DISPID_DTP_PROGRAMMER																	506
#define DISPID_DTP_TESTER																			507
// methods
#define DISPID_DTP_ABOUT																			DISPID_ABOUTBOX
#define DISPID_DTP_CALENDARGETMAXTODAYWIDTH										56
#define DISPID_DTP_CALENDARGETRECTANGLE												57
#define DISPID_DTP_CLOSEDROPDOWNWINDOW												58
#define DISPID_DTP_GETCHECKBOXRECTANGLE												59
#define DISPID_DTP_GETDROPDOWNBUTTONRECTANGLE									60
#define DISPID_DTP_GETIDEALSIZE																61
#define DISPID_DTP_LOADSETTINGSFROMFILE												62
#define DISPID_DTP_OPENDROPDOWNWINDOW													63
#define DISPID_DTP_REFRESH																		DISPID_REFRESH
#define DISPID_DTP_SAVESETTINGSTOFILE													64
#define DISPID_DTP_SETBOLDDATES																65
#define DISPID_DTP_FINISHOLEDRAGDROP													68


// _IDateTimePickerEvents
// methods
#define DISPID_DTPE_CALENDARCLICK															1
#define DISPID_DTPE_CALENDARCLOSEUP														2
#define DISPID_DTPE_CALENDARCONTEXTMENU												3
#define DISPID_DTPE_CALENDARDATEMOUSEENTER										4
#define DISPID_DTPE_CALENDARDATEMOUSELEAVE										5
//#define DISPID_DTPE_CALENDARDBLCLICK													6
#define DISPID_DTPE_CALENDARDROPDOWN													7
#define DISPID_DTPE_CALENDARKEYDOWN														8
#define DISPID_DTPE_CALENDARKEYPRESS													9
#define DISPID_DTPE_CALENDARKEYUP															10
#define DISPID_DTPE_CALENDARMCLICK														11
//#define DISPID_DTPE_CALENDARMDBLCLICK													12
#define DISPID_DTPE_CALENDARMOUSEDOWN													13
#define DISPID_DTPE_CALENDARMOUSEENTER												14
#define DISPID_DTPE_CALENDARMOUSEHOVER												15
#define DISPID_DTPE_CALENDARMOUSELEAVE												16
#define DISPID_DTPE_CALENDARMOUSEMOVE													17
#define DISPID_DTPE_CALENDARMOUSEUP														18
#define DISPID_DTPE_CALENDAROLEDRAGDROP												19
#define DISPID_DTPE_CALENDAROLEDRAGENTER											20
#define DISPID_DTPE_CALENDAROLEDRAGLEAVE											21
#define DISPID_DTPE_CALENDAROLEDRAGMOUSEMOVE									22
#define DISPID_DTPE_CALENDARRCLICK														23
//#define DISPID_DTPE_CALENDARRDBLCLICK													24
#define DISPID_DTPE_CALENDARVIEWCHANGED												25
#define DISPID_DTPE_CALLBACKFIELDKEYDOWN											26
#define DISPID_DTPE_CLICK																			27
#define DISPID_DTPE_CONTEXTMENU																28
#define DISPID_DTPE_CREATEDCALENDARCONTROLWINDOW							29
#define DISPID_DTPE_CREATEDEDITCONTROLWINDOW									30
#define DISPID_DTPE_CURRENTDATECHANGED												31
#define DISPID_DTPE_DBLCLICK																	32
#define DISPID_DTPE_DESTROYEDCALENDARCONTROLWINDOW						33
#define DISPID_DTPE_DESTROYEDCONTROLWINDOW										34
#define DISPID_DTPE_DESTROYEDEDITCONTROLWINDOW								35
#define DISPID_DTPE_GETBOLDDATES															36
#define DISPID_DTPE_GETCALLBACKFIELDTEXT											37
#define DISPID_DTPE_GETCALLBACKFIELDTEXTSIZE									38
#define DISPID_DTPE_KEYDOWN																		39
#define DISPID_DTPE_KEYPRESS																	40
#define DISPID_DTPE_KEYUP																			41
#define DISPID_DTPE_MCLICK																		42
#define DISPID_DTPE_MDBLCLICK																	43
#define DISPID_DTPE_MOUSEDOWN																	44
#define DISPID_DTPE_MOUSEENTER																45
#define DISPID_DTPE_MOUSEHOVER																46
#define DISPID_DTPE_MOUSELEAVE																47
#define DISPID_DTPE_MOUSEMOVE																	48
#define DISPID_DTPE_MOUSEUP																		49
#define DISPID_DTPE_OLEDRAGDROP																50
#define DISPID_DTPE_OLEDRAGENTER															51
#define DISPID_DTPE_OLEDRAGLEAVE															52
#define DISPID_DTPE_OLEDRAGMOUSEMOVE													53
#define DISPID_DTPE_PARSEUSERINPUT														54
#define DISPID_DTPE_RCLICK																		55
#define DISPID_DTPE_RDBLCLICK																	56
#define DISPID_DTPE_RECREATEDCONTROLWINDOW										57
#define DISPID_DTPE_RESIZEDCONTROLWINDOW											58
#define DISPID_DTPE_CALENDARMOUSEWHEEL												59
#define DISPID_DTPE_CALENDARXCLICK														60
#define DISPID_DTPE_MOUSEWHEEL																61
#define DISPID_DTPE_XCLICK																		62
#define DISPID_DTPE_XDBLCLICK																	63


// IOLEDataObject
// properties
// methods
#define DISPID_ODO_CLEAR																			1
#define DISPID_ODO_GETCANONICALFORMAT													2
#define DISPID_ODO_GETDATA																		3
#define DISPID_ODO_GETFORMAT																	4
#define DISPID_ODO_SETDATA																		5
#define DISPID_ODO_GETDROPDESCRIPTION													6
#define DISPID_ODO_SETDROPDESCRIPTION													7
